### PR TITLE
Feature/does pr practice

### PR DESCRIPTION
**Pull request by adelkahomolova**
Implement Does Pull Request Practice

### CODE REVIEW

File: src/inversify.config.ts [hunk 1]
```diff
@@ -41,6 +42,8 @@ export const createRootContainer = (args: ArgumentsProvider): Container => {
   container.bind(FileSystemService).toSelf();
   container.bind(GitHubService).toSelf();
   container.bind(BitbucketService).toSelf();
+  container.bind(Types.ICollaborationInspector).to(CollaborationInspector);
+  container.bind(Types.IIssueTrackingInspector).to(IssueTrackingInspector);
```
Comment (prokopsimek) [inline src/inversify.config.ts]: I guess Inspectors should be bind for a project component, not in a root container

File: src/practices/LanguageIndependent/DoesPullRequests.spec.ts [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    new GitHubNock('1', 'octocat', 1296269, 'Hello-World').getPulls([
+      { number: 1347, state: 'open', title: 'new-feature', body: 'Please pull these awesome changes', head: 'new-topic', base: 'master' },
+    ]);
+    new GitHubNock('1', 'octocat', 1, 'Hello-World').getCommits().reply(200, getRepoCommitsResponse);
+
+    const evaluated = await practice.evaluate(containerCtx.practiceContext);
+    expect(evaluated).toEqual(PracticeEvaluationResult.practicing);
+  });
+
+  it('return notPracticing if there is no PR which is newer than last commit in master minus 30 days', async () => {
```
Comment (prokopsimek) [inline src/practices/LanguageIndependent/DoesPullRequests.spec.ts]: There is no commit? It'd be great to add a test where one commit is really older than -30 days

File: src/practices/LanguageIndependent/DoesPullRequests.ts [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+import { IPractice } from '../IPractice';
+import moment from 'moment';
+
+@DxPractice({
+  id: 'LanguageIndependent.DoesPullRequests',
+  name: 'Do PullRequests',
+  impact: PracticeImpact.medium,
+  suggestion: 'Do pull requests. It helps you catch the bad code before it is merged into the main codebase.',
+  reportOnlyOnce: true,
+  url: 'https://developerexperience.io/practices/pull-requests',
```
Comment (prokopsimek) [inline src/practices/LanguageIndependent/DoesPullRequests.ts]:
```suggestion
  url: 'https://dxkb.io/p/pull-requests',
```

File: src/practices/LanguageIndependent/DoesPullRequests.ts [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+})
+export class DoesPullRequests implements IPractice {
```
Comment (prokopsimek) [inline src/practices/LanguageIndependent/DoesPullRequests.ts]:
```suggestion
export class DoesPullRequestsPractice implements IPractice {
```

Comment (prokopsimek) [reply src/practices/LanguageIndependent/DoesPullRequests.ts]: maybe would be better `DoingPullRequestsPractice`?

File: src/practices/LanguageIndependent/DoesPullRequests.ts [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+    const commitDate = new Date(repoCommits.items[0].author.date).getTime();
+
+    const daysInMilliseconds = moment.duration(30, 'days').asMilliseconds();
+
+    if (prDate > commitDate - daysInMilliseconds) {
+      return PracticeEvaluationResult.practicing;
+    }
+
+    return PracticeEvaluationResult.notPracticing;
```
Comment (prokopsimek) [inline src/practices/LanguageIndependent/DoesPullRequests.ts]:
```
    if (pullRequests.items.length === 0) {
      return PracticeEvaluationResult.notPracticing;
    }

    const prDate = new Date(pullRequests.items[0].createdAt).getTime();
    const commitDate = new Date(repoCommits.items[0].author.date).getTime();
    const daysInMilliseconds = moment.duration(30, 'days').asMilliseconds();

    if (prDate > commitDate - daysInMilliseconds) {
      return PracticeEvaluationResult.practicing;
    }

    return PracticeEvaluationResult.notPracticing;
```
isn't it more readable? :)

File: test/helpers/gitHubNock.ts [hunk 6]
```diff
@@ -190,9 +190,9 @@ export class Repository {
   has_downloads = true;
   archived = false;
   disabled = false;
-  pushed_at = '2011-01-26T19:06:43Z';
-  created_at = '2011-01-26T19:01:12Z';
-  updated_at = '2011-01-26T19:14:43Z';
+  pushed_at = '2012-03-06T23:06:50Z';
+  created_at = '2012-03-06T23:06:50Z';
+  updated_at = '2012-03-06T23:06:50Z';
```
Comment (prokopsimek) [inline test/helpers/gitHubNock.ts]: why?

File: test/helpers/gitHubNock.ts [hunk 7]
```diff
@@ -468,10 +468,10 @@ export class PullRequestItem {
...
   labels = [];
   milestone = null;
-  created_at = '2011-01-26T19:01:12Z';
-  updated_at = '2011-01-26T19:01:12Z';
-  closed_at = '2011-01-26T19:01:12Z';
-  merged_at = '2011-01-26T19:01:12Z';
+  created_at = '2012-03-06T23:06:50Z';
+  updated_at = '2012-03-06T23:06:50Z';
+  closed_at = '2012-03-06T23:06:50Z';
+  merged_at = '2012-03-06T23:06:50Z';
```
Comment (prokopsimek) [inline test/helpers/gitHubNock.ts]: why?

File: src/services/git/__MOCKS__/gitHubServiceMockFolder/getPullServiceResponse.mock.ts [hunk 8]
```diff
@@ -14,12 +14,12 @@ export const getPullServiceResponse: PullRequest = {
...
-  closedAt: '2011-01-26T19:01:12Z',
-  createdAt: '2011-01-26T19:01:12Z',
+  closedAt: '2012-03-06T23:06:50Z',
+  createdAt: '2012-03-06T23:06:50Z',
   id: 1,
-  mergedAt: '2011-01-26T19:01:12Z',
+  mergedAt: '2012-03-06T23:06:50Z',
   state: 'closed',
-  updatedAt: '2011-01-26T19:01:12Z',
+  updatedAt: '2012-03-06T23:06:50Z',
```
Comment (prokopsimek) [inline src/services/git/__MOCKS__/gitHubServiceMockFolder/getPullServiceResponse.mock.ts]: why?

File: src/services/git/__MOCKS__/gitHubServiceMockFolder/getPullsServiceResponse.mock.ts [hunk 9]
```diff
@@ -19,12 +19,12 @@ export const getPullsServiceResponse: Paginated<PullRequest> = {
...
-      closedAt: '2011-01-26T19:01:12Z',
-      createdAt: '2011-01-26T19:01:12Z',
+      closedAt: '2012-03-06T23:06:50Z',
+      createdAt: '2012-03-06T23:06:50Z',
       id: 1,
-      mergedAt: '2011-01-26T19:01:12Z',
+      mergedAt: '2012-03-06T23:06:50Z',
       state: 'open',
-      updatedAt: '2011-01-26T19:01:12Z',
+      updatedAt: '2012-03-06T23:06:50Z',
```
Comment (prokopsimek) [inline src/services/git/__MOCKS__/gitHubServiceMockFolder/getPullsServiceResponse.mock.ts]: why?

File: src/practices/LanguageIndependent/DoesPullRequests.ts [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+    const repoName = GitServiceUtils.getRepoName(ctx.projectComponent.repositoryPath, ctx.projectComponent.path);
+    const ownerAndRepoName = GitServiceUtils.getOwnerAndRepoName(repoName);
+
+    const pullRequests = await ctx.collaborationInspector.getPullRequests(ownerAndRepoName.owner, ownerAndRepoName.repoName);
+    const repoCommits = await ctx.collaborationInspector.getRepoCommits(ownerAndRepoName.owner, ownerAndRepoName.repoName);
+
+    let prDate;
+
+    if (pullRequests.items.length > 0) {
+      prDate = new Date(pullRequests.items[0].createdAt).getTime();
```
Comment (prokopsimek) [inline src/practices/LanguageIndependent/DoesPullRequests.ts]: `pullRequests.items[0]` - We can be sure that the first element is the newest? I'd prefer to use `sort` parameter to ensure the order of pull requests.

File: test/helpers/gitHubNock.ts (hunk 6 above)
Comment (adelkahomolova) [reply test/helpers/gitHubNock.ts]: Just for testing purposes.

File: src/services/git/GitHubService.ts [hunk 11]
```diff
@@ -174,9 +174,35 @@ export class GitHubService implements ICVSService {
...
    * To see the included fields in the verification object see https://octokit.github.io/rest.js/#pagination.
+   *
+   * Sha can be SHA or branch name.
    */
-  async getRepoCommits(owner: string, repo: string) {
-    return this.unwrap(this.client.repos.listCommits({ owner, repo }));
+  async getRepoCommits(owner: string, repo: string, sha?: string): Promise<Paginated<Commit>> {
+    let url = 'GET /repos/:owner/:repo/commits';
+    if (sha !== undefined) {
+      url = `${url}?state=${sha}`;
```
Comment (prokopsimek) [inline src/services/git/GitHubService.ts]: use https://www.npmjs.com/package/qs#stringifying

File: src/services/git/__MOCKS__/gitHubServiceMockFolder/getPullsServiceResponse.mock.ts (hunk 9 above)
Comment (prokopsimek) [reply src/services/git/__MOCKS__/gitHubServiceMockFolder/getPullsServiceResponse.mock.ts]: This is unresolved. Do not mark it as resolved if it's not reverted yet. :)

File: src/services/git/__MOCKS__/gitHubServiceMockFolder/getPullServiceResponse.mock.ts (hunk 8 above)
Comment (prokopsimek) [reply src/services/git/__MOCKS__/gitHubServiceMockFolder/getPullServiceResponse.mock.ts]: This is unresolved. Do not mark it as resolved if it's not reverted yet. :)

File: test/helpers/gitHubNock.ts [hunk 12]
```diff
@@ -26,26 +26,78 @@ export class GitHubNock {
...
+      updated_at?: string;
+    }[],
     queryState?: string,
     persist = true,
   ): PullRequestItem[] {
-    const responseBody = pulls.map(
-      ({ number, state, title, body, head, base }) =>
-        new PullRequestItem(number, state, title, body, new BranchItem(head, this.repository), new BranchItem(base, this.repository)),
-    );
+    // eslint-disable-next-line @typescript-eslint/camelcase
```
Comment (prokopsimek) [inline test/helpers/gitHubNock.ts]: use the disable for whole file instead of this messy lines

File: test/helpers/gitHubNock.ts (hunk 6 above)
Comment (prokopsimek) [reply test/helpers/gitHubNock.ts]: This is unresolved. Do not mark it as resolved if it's not reverted yet. :)

File: test/helpers/gitHubNock.ts (hunk 7 above)
Comment (prokopsimek) [reply test/helpers/gitHubNock.ts]: This is unresolved. Do not mark it as resolved if it's not reverted yet. :)

File: src/services/git/GitHubService.ts (hunk 11 above)
Comment (adelkahomolova) [reply src/services/git/GitHubService.ts]: Fixed here https://github.com/DXHeroes/dx-scanner/pull/164/files#diff-e8b7dc86926ce489d3e57e63817c7979R79

Comment (prokopsimek) [reply src/services/git/GitHubService.ts]: not yet merged `master`?

**Pull request merged**